### PR TITLE
use FadeTransition instead of Opacity where applicable

### LINKED ISSUE #75109: review framework to use FadeTransition instead of Opacity where applicable
according to https://flutter.dev/docs/perf/rendering/best-practices it's better to use `FadeTransition` for animations, but in the framework itself there's a few places where this is not applied and `Opacity` is used isntead of `FadeTransition`

### PR DESCRIPTION
**Pull request by nt4f04uNd**
this PR focuses on chaning core framework widgets that in some places misuse the `Opacity` inside animations

fixes https://github.com/flutter/flutter/issues/75109

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by nt4f04uNd** [comment]
i commited the changes only to the core
there are a few other places in other different places like tests that i didn't touch

also, there are a few places that i doubt if they should be replaced, because in there i have to actually instantiate an `AnimationContoller` to pass it to `FadeTransition`. i'm gonna investigate how it is better to handle this and open another PR, so this one is not in progress

**Comment by nt4f04uNd** [comment]
@goderbauer as i mentioned in https://github.com/flutter/flutter/pull/75110#issuecomment-770528917, there's a few places where i doubt what has to be done

* https://github.com/flutter/flutter/blob/2c5c50948686fb698842317c3f5d2bf54c6090ca/packages/flutter/lib/src/material/app_bar.dart#L910-L913
* https://github.com/flutter/flutter/blob/2c5c50948686fb698842317c3f5d2bf54c6090ca/packages/flutter/lib/src/cupertino/refresh.dart#L454-L457
* https://github.com/flutter/flutter/blob/2c5c50948686fb698842317c3f5d2bf54c6090ca/packages/flutter/lib/src/material/flexible_space_bar.dart#L314-L320
* https://github.com/flutter/flutter/blob/2c5c50948686fb698842317c3f5d2bf54c6090ca/packages/flutter/lib/src/material/flexible_space_bar.dart#L367-L370

FadeTransition (and RenderAnimatedOpacityMixin) allows only to use an Animation, but there we might want to use just a raw value for them. now to do that you have to additionally instantiate a State with AnimationController, which is pretty expensive i guess

* so is there any particular benchmarks comparing the usage of Opacity and FadeTransition for animations?
* why Opacity is not optimized (and just considered to be expensive instead of being optimized) for these purposes in the first place? what's the tradeoff between them i mean
* if it's reasonable to keep both Opacity and FadeTransition, and not optimize the former one, probably we should add something like `FadeTransition.transform`/`FadeTransition.fromValue` to make it enabled for uses

**Comment by goderbauer** [comment]
This sounds good. However, before I review this in detail: It looks like there are some failing tests. Can you take a look at those?

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -441,8 +441,8 @@ class _HelperErrorState extends State<_HelperError> with SingleTickerProviderSta
     if (widget.errorText != null) {
       return Stack(
         children: <Widget>[
-          Opacity(
-            opacity: 1.0 - _controller.value,
+          FadeTransition(
+            opacity: Tween<double>(begin: 1.0, end: 0.0).animate(_controller),
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/input_decorator.dart:445]: Above you use drive, here animate. Can you pick one?

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 2]
```diff
@@ -2316,14 +2316,28 @@ Finder _opacityAboveLabel(String text) {
...
-    find.ancestor(
-      of: find.text(text),
-      matching: find.byType(Opacity),
-    ),
-  );
-  return opacityWidget.opacity;
+double? _labelOpacity(WidgetTester tester, String text) {
+  // We search for both Opacity and FadeTransition since in some
+  // cases opacity is animated, in other it's not.
+  try {
```
Comment (goderbauer) [inline packages/flutter/test/material/navigation_rail_test.dart:2322]: please don't use try/catch for control flow.

File: packages/flutter/lib/src/material/navigation_rail.dart [hunk 3]
```diff
@@ -667,9 +668,9 @@ class _RailDestination extends StatelessWidget {
                       heightFactor: 1.0,
                       widthFactor: extendedTransitionAnimation.value,
                       alignment: AlignmentDirectional.centerStart,
-                      child: Opacity(
+                      child: FadeTransition(
                         alwaysIncludeSemantics: true,
-                        opacity: _extendedLabelFadeValue(),
+                        opacity: labelFadeAnimation,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/navigation_rail.dart:673]: This is odd: previously, the animation was driven by extendedTransitionAnimation, now it is destinationAnimation. Why this change?

**Comment by goderbauer** [comment]
@nt4f04uNd Do you still have plans to come back to this?

**Comment by nt4f04uNd** [comment]
yes, i will update it soon

File: packages/flutter/lib/src/material/navigation_rail.dart (hunk 3 above)
Comment (nt4f04uNd) [reply packages/flutter/lib/src/material/navigation_rail.dart:673]: copy-paste mistake, fixed

File: packages/flutter/test/material/navigation_rail_test.dart (hunk 2 above)
Comment (nt4f04uNd) [reply packages/flutter/test/material/navigation_rail_test.dart:2322]: we both use Opacity and FadeTransition - if former was not found in tree, it should pick the latter one. i am not sure what i should do about this?

Comment (nt4f04uNd) [reply packages/flutter/test/material/navigation_rail_test.dart]: fixed it

**Comment by nt4f04uNd** [comment]
@goderbauer i believe it's ready for another review. it looks like some tests still failed due to reasons unrelated to this pr

**Comment by goderbauer** [comment]
@nt4f04uNd Can you rebase this to the latest master to get the tests to pass? Thanks!

**Comment by nt4f04uNd** [comment]
@goderbauer hi, is there anything else I'd need to do to get this merged?

**Comment by goderbauer** [comment]
This change LGTM, but we are currently working on adjusting the internals of FadeTransition in https://github.com/flutter/flutter/pull/83145. Since that's requiring some golden updates, I'd like to get that one in first before using FadeTransition for more things.

**Comment by Piinks** [comment]
Hey @nt4f04uNd, it looks like #83145 has landed. Can you update this PR with the latest from master, I think this is ready to land. :)

**Comment by nt4f04uNd** [comment]
Hi, would you mind merging this?

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 4]
```diff
@@ -135,8 +135,7 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
   /// Indicates whether the tab bar is fully opaque or can have contents behind
   /// it show through it.
   bool opaque(BuildContext context) {
-    final Color backgroundColor =
-        this.backgroundColor ?? CupertinoTheme.of(context).barBackgroundColor;
+    final Color backgroundColor = this.backgroundColor ?? CupertinoTheme.of(context).barBackgroundColor;
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:138]: Can you revert the formatting changes in this file?

File: packages/flutter/lib/src/cupertino/activity_indicator.dart [hunk 5]
```diff
@@ -117,8 +117,7 @@ class _CupertinoActivityIndicatorState extends State<CupertinoActivityIndicator>
       child: CustomPaint(
         painter: _CupertinoActivityIndicatorPainter(
           position: _controller,
-          activeColor:
-              CupertinoDynamicColor.resolve(_kActiveTickColor, context),
+          activeColor: CupertinoDynamicColor.resolve(_kActiveTickColor, context),
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:120]: Same for formatting changes in this file.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you!

**Comment by Piinks** [comment]
I'll monitor this in case the flutter-gold check prevents it from merging. The fix for that is still blocked by https://github.com/flutter/flutter/issues/89709
@nt4f04uNd if you notice the other checks complete before I do, feel free to ping here.

**Comment by nt4f04uNd** [comment]
Looks like it's all green now
